File: oracle/go/relayer/relayer.go
package relayer

import (
	"context"
	"fmt"
	"strings"
	"time"

	rpchttp "github.com/cometbft/cometbft/rpc/client/http"

	"github.com/cosmos/cosmos-sdk/client"
	sdk "github.com/cosmos/cosmos-sdk/types"
	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"

	oracledatatypes "steemvm/x/oracle/data/types"
	"steemvm/x/oracle/bridge/types"
)

// Run drives the Steem relayer loop until ctx is canceled. It is a pure
// library entry point: the standalone oracle binary supplies a fully-built
// clientCtx (codec + in-memory keyring holding cfg.KeyName), the SteemVM node
// RPC endpoint to broadcast/query against, the relayer Config, and a directory
// for the scan-cursor state file. Every cycle failure is logged and retried;
// Run only returns on ctx cancellation or a fatal setup error.
//
// priceSource, if non-nil, activates the price feeder (x/oracle/data) on the
// same poll loop: gasPrices must be set too (price-feed txs are NOT
// fee-exempt — see oracle/PROTOCOL.md §3). A nil priceSource idles the
// feeder entirely, same as today's default.
func Run(ctx context.Context, logger cycleLogger, clientCtx client.Context, nodeRPC string, cfg Config, stateDir string, priceSource PriceSource, gasPrices string) error {
	// Resolve the signing key from the supplied keyring.
	if clientCtx.Keyring == nil {
		return fmt.Errorf("no keyring available on client context")
	}
	record, err := clientCtx.Keyring.Key(cfg.KeyName)
	if err != nil {
		return fmt.Errorf("key %q not found in keyring: %w", cfg.KeyName, err)
	}
	signerAddr, err := record.GetAddress()
	if err != nil {
		return fmt.Errorf("cannot read key address: %w", err)
	}
	valoperAddr := sdk.ValAddress(signerAddr).String()

	// Connect to the SteemVM node RPC (broadcast + ABCI queries route through it).
	rpcClient, err := rpchttp.New(nodeRPC, "/websocket")
	if err != nil {
		return fmt.Errorf("cannot create node RPC client for %q: %w", nodeRPC, err)
	}
	clientCtx = clientCtx.
		WithClient(rpcClient).
		WithFromName(cfg.KeyName).
		WithFromAddress(signerAddr).
		WithBroadcastMode("sync")

	// Wait until the node is up and synced, then learn the chain-id.
	chainID := waitForLocalNode(ctx, logger, rpcClient)
	if chainID == "" {
		return ctx.Err() // ctx canceled
	}
	clientCtx = clientCtx.WithChainID(chainID)

	steem := NewSteemClient(cfg.SteemRPCURL)
	bridgeQuery := types.NewQueryClient(clientCtx)
	stakingQuery := stakingtypes.NewQueryClient(clientCtx)
	oracledataQuery := oracledatatypes.NewQueryClient(clientCtx)

	state, err := LoadState(stateDir)
	if err != nil {
		return fmt.Errorf("cannot load state from %q: %w", stateDir, err)
	}

	feeder := Feeder{Validator: signerAddr.String(), Source: priceSource}
	// lastHandledPeriod guards against re-firing Step (and re-paying gas) more
	// than once per vote period across ticks; 0 is a safe "never handled"
	// sentinel in practice (a chain's genesis-era period 0 sees at most a
	// handful of redundant prevotes, harmless).
	var lastHandledPeriod uint64

	logger.Info("steem oracle started",
		"steem_rpc", cfg.SteemRPCURL, "node_rpc", nodeRPC, "chain_id", chainID,
		"signer", signerAddr.String(), "valoper", valoperAddr,
		"poll_interval", cfg.PollInterval.String(), "last_scanned_block", state.LastScannedBlock,
		"price_feeder_enabled", priceSource != nil)

	ticker := time.NewTicker(cfg.PollInterval)
	defer ticker.Stop()
	notBondedLogged := false

	for {
		select {
		case <-ctx.Done():
			logger.Info("steem oracle stopped")
			return nil
		case <-ticker.C:
		}

		if err := runCycle(ctx, logger, cfg, clientCtx, steem, bridgeQuery, stakingQuery, signerAddr, valoperAddr, stateDir, &state, &notBondedLogged); err != nil {
			logger.Error("steem oracle cycle failed", "err", err)
		}

		if priceSource != nil {
			if err := runPriceFeederCycle(ctx, logger, clientCtx, oracledataQuery, feeder, gasPrices, stateDir, &lastHandledPeriod); err != nil {
				logger.Error("price feeder cycle failed", "err", err)
			}
		}
	}
}

// runPriceFeederCycle checks whether the chain has entered a new vote period
// since the last handled one and, if so, runs one Feeder.Step: reveals the
// prior period's commit (if still in the reveal window) and/or commits a
// fresh prevote, broadcasting whatever messages result and persisting the
// new commit state. A no-op (same period as last time) costs nothing.
func runPriceFeederCycle(
	ctx context.Context,
	logger cycleLogger,
	clientCtx client.Context,
	oracledataQuery oracledatatypes.QueryClient,
	feeder Feeder,
	gasPrices string,
	stateDir string,
	lastHandledPeriod *uint64,
) error {
	paramsResp, err := oracledataQuery.Params(ctx, &oracledatatypes.QueryParamsRequest{})
	if err != nil {
		return fmt.Errorf("querying oracledata params: %w", err)
	}
	votePeriod := paramsResp.Params.VotePeriod
	if votePeriod == 0 {
		return fmt.Errorf("oracledata vote period is zero")
	}

	status, err := clientCtx.Client.Status(ctx)
	if err != nil {
		return fmt.Errorf("querying node status: %w", err)
	}
	height := uint64(status.SyncInfo.LatestBlockHeight) //nolint:gosec // block heights never approach int64 overflow
	period := height / votePeriod

	if period == *lastHandledPeriod {
		return nil // already acted this period; wait for the next boundary
	}

	prevState, err := LoadFeederState(stateDir)
	if err != nil {
		return fmt.Errorf("cannot load price feeder state from %q: %w", stateDir, err)
	}

	msgs, newState, err := feeder.Step(period, paramsResp.Params.Whitelist, prevState)
	if err != nil {
		// A price-fetch error still lets msgs (a pending reveal, if any)
		// through — only the fresh-prevote half of Step failed.
		logger.Error("price feeder: fetching prices failed", "err", err)
	}
	if len(msgs) == 0 {
		*lastHandledPeriod = period
		return nil
	}

	txHash, err := BroadcastPriceFeedMsgs(ctx, clientCtx, clientCtx.FromName, msgs, gasPrices)
	if err != nil {
		return fmt.Errorf("broadcasting price feed messages: %w", err)
	}
	logger.Info("price feeder broadcast", "period", period, "count", len(msgs), "tx", txHash)

	if err := SaveFeederState(stateDir, newState); err != nil {
		return fmt.Errorf("saving price feeder state: %w", err)
	}
	*lastHandledPeriod = period
	return nil
}

type cycleLogger interface {
	Info(msg string, keyVals ...any)
	Error(msg string, keyVals ...any)
	Debug(msg string, keyVals ...any)
}

// runCycle performs one poll: check feature flags and bonded status, scan
// new irreversible Steem blocks, attest gateway transfers, advance cursor.
func runCycle(
	ctx context.Context,
	logger cycleLogger,
	cfg Config,
	clientCtx client.Context,
	steem *SteemClient,
	bridgeQuery types.QueryClient,
	stakingQuery stakingtypes.QueryClient,
	signerAddr sdk.AccAddress,
	valoperAddr string,
	stateDir string,
	state *State,
	notBondedLogged *bool,
) error {
	params, err := bridgeQuery.Params(ctx, &types.QueryParamsRequest{})
	if err != nil {
		return err
	}
	// GatewayAccount is a hardcoded chain constant, never read from params
	// (see x/oracle/bridge/types.GatewayAccount) — this closes off the class
	// of mistake where a stale/wrong configured value silently diverges from
	// what the chain actually enforces.
	gateway := types.GatewayAccount
	if !params.Params.BridgeEnabled && !params.Params.NameServiceEnabled {
		logger.Debug("steem relayer idle: bridge and name service disabled")
		return nil
	}

	// Only bonded validators' attestations count (and only theirs are
	// fee-exempt) — idle quietly otherwise.
	valResp, err := stakingQuery.Validator(ctx, &stakingtypes.QueryValidatorRequest{ValidatorAddr: valoperAddr})
	if err != nil || !valResp.Validator.IsBonded() {
		if !*notBondedLogged {
			logger.Info("steem relayer idle: key is not a bonded validator", "valoper", valoperAddr)
			*notBondedLogged = true
		}
		return nil
	}
	*notBondedLogged = false

	lib, err := steem.LastIrreversibleBlock()
	if err != nil {
		return err
	}

	// First run: establish the cursor. Precedence: the node's local
	// start-block override, then the chain-wide relayer_start_block param
	// (the shared anchor recorded at launch so every validator scans the
	// same range regardless of when it first starts), then the current LIB.
	if state.LastScannedBlock == 0 {
		state.LastScannedBlock = initialCursor(cfg.StartBlock, params.Params.RelayerStartBlock, lib)
		logger.Info("steem relayer cursor initialized",
			"cursor", state.LastScannedBlock,
			"source", cursorSource(cfg.StartBlock, params.Params.RelayerStartBlock))
		return SaveState(stateDir, *state)
	}

	from := state.LastScannedBlock + 1
	to := min(lib, state.LastScannedBlock+cfg.MaxBlocksPerPoll)
	if from > to {
		return nil
	}

	blocks, err := steem.FetchBlocks(from, to)
	if err != nil {
		return err
	}

	// Accumulate attestations across blocks, keeping block boundaries so the
	// cursor only advances past fully-included blocks. One tx per cycle.
	var msgs []sdk.Msg
	lastFullBlock := state.LastScannedBlock
	for _, nb := range blocks {
		transfers := ExtractGatewayTransfers(nb.Num, nb.Block, gateway, cfg.SteemSymbol, cfg.SbdSymbol)
		blockMsgs := make([]sdk.Msg, 0, len(transfers))
		for _, transfer := range transfers {
			intent := RouteMemo(transfer.Memo)
			if intent == IntentRegister && !params.Params.NameServiceEnabled {
				continue
			}
			if intent == IntentDeposit && !params.Params.BridgeEnabled {
				continue
			}
			// Only attest transfers whose memo resolves to a supported
			// destination (a steem1.../0x address, optionally svm-deposit /
			// svm-register prefixed) — the exact in-consensus parser. Transfers
			// with an unparseable memo (a human note, a wrong address, etc.) are
			// ignored: we don't broadcast attestations, and no on-chain record
			// is created for transfers the bridge cannot act on.
			if _, _, ok := types.DeriveDestination(transfer.Memo); !ok {
				continue
			}
			attested, err := alreadyAttested(ctx, bridgeQuery, transfer, intent, valoperAddr)
			if err != nil {
				return err
			}
			if attested {
				continue
			}
			blockMsgs = append(blockMsgs, BuildMsg(transfer, intent, signerAddr, gateway))
		}
		// Withdrawal payouts: gateway→user transfers memoed "svm-withdrawal <id>".
		// Optimistic — every payout seen is attested once (the cursor scans each
		// block exactly once); the chain benign-no-ops duplicates and finalizes the
		// withdrawal REQUESTED→PROCESSED at 2/3, so no per-validator dedup query is
		// needed here. Gated on bridge-out being enabled (the v0.0.3 payout path).
		if params.Params.BridgeOutEnabled {
			for _, payout := range ExtractGatewayPayouts(nb.Num, nb.Block, gateway) {
				blockMsgs = append(blockMsgs, &types.MsgAttestWithdrawalPayout{
					Validator:      signerAddr.String(),
					WithdrawalId:   payout.WithdrawalID,
					SteemTxid:      payout.Txid,
					OpIndex:        payout.OpIndex,
					SteemBlock:     payout.SteemBlock,
					SteemTimestamp: payout.SteemTimestamp,
				})
			}
		}
		if len(msgs)+len(blockMsgs) > MaxMsgsPerTx {
			break // stop at a block boundary; the rest is picked up next cycle
		}
		msgs = append(msgs, blockMsgs...)
		lastFullBlock = nb.Num
	}

	if len(msgs) > 0 {
		txHash, err := BroadcastAttestations(ctx, clientCtx, cfg.KeyName, msgs)
		if err != nil {
			return err // cursor untouched; the same blocks are rescanned next cycle
		}
		logger.Info("steem relayer attested transfers",
			"count", len(msgs), "blocks", lastFullBlock-state.LastScannedBlock, "tx", txHash)
	}

	if lastFullBlock != state.LastScannedBlock {
		state.LastScannedBlock = lastFullBlock
		return SaveState(stateDir, *state)
	}
	return nil
}

// initialCursor picks the first-run scan cursor (the last "already scanned"
// block, i.e. scanning begins at cursor+1): a local app.toml start-block
// override wins, then the chain's relayer_start_block param, then Steem's
// current last irreversible block.
func initialCursor(localStart, chainStart, lib uint64) uint64 {
	switch {
	case localStart > 0:
		return localStart - 1
	case chainStart > 0:
		return chainStart - 1
	default:
		return lib
	}
}

func cursorSource(localStart, chainStart uint64) string {
	switch {
	case localStart > 0:
		return "app.toml start-block override"
	case chainStart > 0:
		return "on-chain relayer_start_block param"
	default:
		return "current steem last irreversible block"
	}
}

// alreadyAttested reports whether this validator has already confirmed the
// transfer's (txid, opIndex) key, or the record is past the point where new
// attestations are accepted. Keeps rescans (lost state file, retried cycles)
// idempotent — a duplicate message would disqualify the whole zero-fee tx.
func alreadyAttested(ctx context.Context, q types.QueryClient, t Transfer, intent Intent, valoperAddr string) (bool, error) {
	switch intent {
	case IntentRegister:
		resp, err := q.NameRegistrationByTxid(ctx, &types.QueryNameRegistrationByTxidRequest{Txid: t.Txid, OpIndex: t.OpIndex})
		if err != nil {
			return false, ignoreNotFound(err)
		}
		if resp.Registration.Status != types.NameRegistrationStatus_NAME_REGISTRATION_STATUS_PENDING {
			return true, nil
		}
		return hasConfirmation(resp.Registration.ValidatorConfirmations, valoperAddr), nil
	default:
		resp, err := q.DepositByTxid(ctx, &types.QueryDepositByTxidRequest{Txid: t.Txid, OpIndex: t.OpIndex})
		if err != nil {
			return false, ignoreNotFound(err)
		}
		if resp.Deposit.Status != types.DepositStatus_DEPOSIT_STATUS_PENDING {
			return true, nil
		}
		return hasConfirmation(resp.Deposit.ValidatorConfirmations, valoperAddr), nil
	}
}

func hasConfirmation(confirmations []*types.Confirmation, valoperAddr string) bool {
	for _, c := range confirmations {
		if c.ValidatorAddress == valoperAddr {
			return true
		}
	}
	return false
}

// ignoreNotFound maps "key not found" query errors to nil (meaning: not yet
// attested); anything else is a real error that aborts the cycle. ABCI query
// errors arrive as flattened strings, so this matches on the message.
func ignoreNotFound(err error) error {
	if err == nil {
		return nil
	}
	if strings.Contains(err.Error(), "not found") || strings.Contains(err.Error(), "NotFound") {
		return nil
	}
	return err
}

// waitForLocalNode blocks until the local node RPC responds and is done
// catching up, returning the chain-id (empty when ctx is canceled).
func waitForLocalNode(ctx context.Context, logger cycleLogger, rpcClient *rpchttp.HTTP) string {
	for {
		select {
		case <-ctx.Done():
			return ""
		case <-time.After(2 * time.Second):
		}
		status, err := rpcClient.Status(ctx)
		if err != nil {
			continue // node still booting
		}
		if status.SyncInfo.CatchingUp {
			logger.Debug("steem relayer waiting: node is catching up")
			continue
		}
		return status.NodeInfo.Network
	}
}

File: oracle/python/src/steemvm_oracle/relayer.py
"""Main scan/attest loop. Ports ``oracle/go/relayer/relayer.go``, adapted
for this client's REST-only transport (``oracle/PROTOCOL.md`` SS4): every
CometBFT-RPC/gRPC query the Go client makes has a REST (gRPC-gateway)
equivalent used here instead.

Bridge attestations are simpler than the price feeder (no fee/gas-price
concerns -- fee-exempt for bonded validators), so this module drives both:
one bridge-scan cycle, then (if configured) one price-feeder cycle, per
poll tick -- matching the Go reference's combined loop.
"""

from __future__ import annotations

import logging
import threading
from dataclasses import dataclass
from typing import Optional

from . import _protopath  # noqa: F401
from .broadcast import MAX_MSGS_PER_TX, NotFoundError, RestClient, broadcast_attestations, broadcast_price_feed_msgs
from .config import Config
from .keys import Keypair
from .pricefeeder import CompositePriceSource, Feeder
from .router import GATEWAY_ACCOUNT, Intent, build_payout_any, build_transfer_any, derive_destination, route_memo
from .state import FeederState, State, load_feeder_state, load_state, save_feeder_state, save_state
from .steem_client import SteemClient, extract_gateway_payouts, extract_gateway_transfers

logger = logging.getLogger("steemvm_oracle.relayer")


# --------------------------------------------------------------------------
# Node queries (REST). Every one has a CometBFT-RPC/gRPC counterpart in the
# Go client; oracle/PROTOCOL.md SS4 explains why REST is preferred here.
# --------------------------------------------------------------------------


def query_bridge_params(rest: RestClient) -> dict:
    return rest.get_json("/steemvm/steembridge/v1/params")["params"]


def query_oracledata_params(rest: RestClient) -> dict:
    return rest.get_json("/steemvm/oracle/data/v1/params")["params"]


def query_validator_bonded(rest: RestClient, valoper_address: str) -> bool:
    try:
        resp = rest.get_json(f"/cosmos/staking/v1beta1/validators/{valoper_address}")
    except NotFoundError:
        return False
    return resp.get("validator", {}).get("status") == "BOND_STATUS_BONDED"


def query_deposit_by_txid(rest: RestClient, txid: str, op_index: int) -> Optional[dict]:
    try:
        return rest.get_json(f"/steemvm/steembridge/v1/deposit_by_txid/{txid}/{op_index}")["deposit"]
    except NotFoundError:
        return None


def query_name_registration_by_txid(rest: RestClient, txid: str, op_index: int) -> Optional[dict]:
    try:
        return rest.get_json(f"/steemvm/steembridge/v1/name_registration_by_txid/{txid}/{op_index}")[
            "registration"
        ]
    except NotFoundError:
        return None


def query_node_chain_id(rest: RestClient) -> str:
    return rest.get_json("/cosmos/base/tendermint/v1beta1/node_info")["default_node_info"]["network"]


def query_is_syncing(rest: RestClient) -> bool:
    return bool(rest.get_json("/cosmos/base/tendermint/v1beta1/syncing").get("syncing", False))


def query_latest_height(rest: RestClient) -> int:
    resp = rest.get_json("/cosmos/base/tendermint/v1beta1/blocks/latest")
    return int(resp["block"]["header"]["height"])


# --------------------------------------------------------------------------
# Cursor / attestation-dedup helpers
# --------------------------------------------------------------------------


def _initial_cursor(local_start: int, chain_start: int, lib: int) -> int:
    """Picks the first-run scan cursor (the last "already scanned" block,
    i.e. scanning begins at cursor+1). Precedence: a local start-block
    override, then the chain-wide relayer_start_block param, then Steem's
    current last irreversible block."""
    if local_start > 0:
        return local_start - 1
    if chain_start > 0:
        return chain_start - 1
    return lib


def _has_confirmation(confirmations: list[dict], valoper_address: str) -> bool:
    return any(c.get("validator_address") == valoper_address for c in confirmations)


def _already_attested(rest: RestClient, transfer, intent: Intent, valoper_address: str) -> bool:
    """Reports whether this validator has already confirmed the transfer's
    (txid, op_index) key. Keeps rescans idempotent -- a duplicate message
    would disqualify the whole zero-fee tx. Mirrors
    oracle/go/relayer/relayer.go's alreadyAttested."""
    if intent == Intent.REGISTER:
        reg = query_name_registration_by_txid(rest, transfer.txid, transfer.op_index)
        if reg is None:
            return False
        if reg.get("status") != "NAME_REGISTRATION_STATUS_PENDING":
            return True
        return _has_confirmation(reg.get("validator_confirmations", []), valoper_address)

    dep = query_deposit_by_txid(rest, transfer.txid, transfer.op_index)
    if dep is None:
        return False
    if dep.get("status") != "DEPOSIT_STATUS_PENDING":
        return True
    return _has_confirmation(dep.get("validator_confirmations", []), valoper_address)


# --------------------------------------------------------------------------
# Poll cycles
# --------------------------------------------------------------------------


@dataclass
class Cycle:
    """Long-lived handles a poll cycle needs."""

    rest: RestClient
    steem: SteemClient
    keypair: Keypair
    chain_id: str
    cfg: Config
    state_dir: str


def run_cycle(cycle: Cycle, state: State, not_bonded_logged: list[bool]) -> State:
    """One poll: check feature flags and bonded status, scan new
    irreversible Steem blocks, attest gateway transfers, advance cursor.
    Mirrors oracle/go/relayer/relayer.go's runCycle."""
    params = query_bridge_params(cycle.rest)
    # GatewayAccount is a hardcoded chain constant, never read from params --
    # see router.GATEWAY_ACCOUNT's doc comment.
    gateway = GATEWAY_ACCOUNT
    bridge_enabled = bool(params.get("bridge_enabled"))
    name_service_enabled = bool(params.get("name_service_enabled"))
    bridge_out_enabled = bool(params.get("bridge_out_enabled"))

    if not (bridge_enabled or name_service_enabled):
        logger.debug("steem relayer idle: bridge and name service disabled")
        return state

    # Only bonded validators' attestations count (and only theirs are
    # fee-exempt) -- idle quietly otherwise.
    if not query_validator_bonded(cycle.rest, cycle.keypair.valoper_address):
        if not not_bonded_logged[0]:
            logger.info("steem relayer idle: key is not a bonded validator (valoper=%s)", cycle.keypair.valoper_address)
            not_bonded_logged[0] = True
        return state
    not_bonded_logged[0] = False

    lib = cycle.steem.last_irreversible_block()

    # First run: establish the cursor.
    if state.last_scanned_block == 0:
        state.last_scanned_block = _initial_cursor(cycle.cfg.start_block, int(params.get("relayer_start_block", 0)), lib)
        logger.info("steem relayer cursor initialized: cursor=%d", state.last_scanned_block)
        save_state(cycle.state_dir, state)
        return state

    from_block = state.last_scanned_block + 1
    to_block = min(lib, state.last_scanned_block + cycle.cfg.max_blocks_per_poll)
    if from_block > to_block:
        return state

    blocks = cycle.steem.fetch_blocks(from_block, to_block)

    # Accumulate attestations across blocks, keeping block boundaries so the
    # cursor only advances past fully-included blocks. One tx per cycle.
    msgs = []
    last_full_block = state.last_scanned_block
    for block_num, block in blocks:
        transfers = extract_gateway_transfers(block_num, block, gateway, cycle.cfg.steem_symbol, cycle.cfg.sbd_symbol)
        block_msgs = []
        for transfer in transfers:
            intent = route_memo(transfer.memo)
            if intent == Intent.REGISTER and not name_service_enabled:
                continue
            if intent == Intent.DEPOSIT and not bridge_enabled:
                continue
            # Only attest transfers whose memo resolves to a supported
            # destination -- the exact in-consensus parser.
            _, _, ok = derive_destination(transfer.memo)
            if not ok:
                continue
            if _already_attested(cycle.rest, transfer, intent, cycle.keypair.valoper_address):
                continue
            block_msgs.append(build_transfer_any(transfer, intent, cycle.keypair.address, gateway))
            intent_label = "name-registration" if intent == Intent.REGISTER else "deposit"
            logger.info(
                "attesting transfer: intent=%s txid=%s op_index=%d from=%s amount_millisteem=%d memo=%s",
                intent_label, transfer.txid, transfer.op_index, transfer.from_,
                transfer.amount_millisteem, transfer.memo,
            )

        # Withdrawal payouts: optimistic, no per-validator dedup query
        # needed (chain benign-no-ops duplicates at 2/3 threshold).
        if bridge_out_enabled:
            for payout in extract_gateway_payouts(block_num, block, gateway, cycle.cfg.steem_symbol, cycle.cfg.sbd_symbol):
                block_msgs.append(build_payout_any(payout, cycle.keypair.address))
                logger.info(
                    "attesting withdrawal payout: withdrawal_id=%d steem_txid=%s op_index=%d",
                    payout.withdrawal_id, payout.txid, payout.op_index,
                )

        if len(msgs) + len(block_msgs) > MAX_MSGS_PER_TX:
            break  # stop at a block boundary; the rest is picked up next cycle
        msgs.extend(block_msgs)
        last_full_block = block_num

    if msgs:
        tx_hash = broadcast_attestations(cycle.rest, cycle.keypair, cycle.chain_id, msgs)
        logger.info(
            "steem relayer attested transfers: count=%d blocks=%d tx=%s",
            len(msgs), last_full_block - state.last_scanned_block, tx_hash,
        )

    if last_full_block != state.last_scanned_block:
        state.last_scanned_block = last_full_block
        save_state(cycle.state_dir, state)

    return state


def run_price_feeder_cycle(
    cycle: Cycle,
    feeder: Feeder,
    gas_prices: str,
    last_handled_period: list[int],
) -> None:
    """Checks whether the chain has entered a new vote period since the
    last handled one and, if so, runs one Feeder.step. Mirrors
    oracle/go/relayer/relayer.go's runPriceFeederCycle."""
    oracle_params = query_oracledata_params(cycle.rest)
    vote_period = int(oracle_params.get("vote_period", 0))
    if vote_period == 0:
        raise RuntimeError("oracledata vote period is zero")
    whitelist = oracle_params.get("whitelist", [])

    height = query_latest_height(cycle.rest)
    period = height // vote_period

    if period == last_handled_period[0]:
        return  # already acted this period; wait for the next boundary

    prev_state: FeederState = load_feeder_state(cycle.state_dir)
    msgs, new_state = feeder.step(period, whitelist, prev_state)

    if not msgs:
        last_handled_period[0] = period
        return

    tx_hash = broadcast_price_feed_msgs(cycle.rest, cycle.keypair, cycle.chain_id, msgs, gas_prices)
    logger.info("price feeder broadcast: period=%d count=%d tx=%s", period, len(msgs), tx_hash)

    save_feeder_state(cycle.state_dir, new_state)
    last_handled_period[0] = period


# --------------------------------------------------------------------------
# Top-level Run loop
# --------------------------------------------------------------------------


def wait_for_local_node(rest: RestClient, stop: threading.Event) -> str:
    """Blocks until the local node REST responds and is done catching up,
    returning the chain-id (empty when stop is set). Mirrors
    oracle/go/relayer/relayer.go's waitForLocalNode."""
    while not stop.is_set():
        if stop.wait(2.0):
            return ""
        try:
            if query_is_syncing(rest):
                logger.debug("steem relayer waiting: node is catching up")
                continue
            return query_node_chain_id(rest)
        except Exception:  # noqa: BLE001 - node still booting
            continue
    return ""


def run(
    keypair: Keypair,
    node_rest: str,
    cfg: Config,
    state_dir: str,
    price_source: Optional[CompositePriceSource],
    gas_prices: str,
    stop: threading.Event,
) -> None:
    """Drives the relayer loop until `stop` is set. Mirrors
    oracle/go/relayer/relayer.go's Run. `price_source`, if non-None,
    activates the price feeder (x/oracle/data) on the same poll loop --
    gas_prices must be set too (price-feed txs are NOT fee-exempt)."""
    rest = RestClient(node_rest)
    steem = SteemClient(cfg.steem_rpc_url)
    try:
        chain_id = wait_for_local_node(rest, stop)
        if not chain_id:
            return  # stop was set while waiting

        state = load_state(state_dir)
        feeder = Feeder(validator=keypair.address, source=price_source)
        last_handled_period = [0]
        not_bonded_logged = [False]

        logger.info(
            "steem oracle started: steem_rpc=%s node_rest=%s chain_id=%s signer=%s valoper=%s "
            "poll_interval=%ss last_scanned_block=%d price_feeder_enabled=%s",
            cfg.steem_rpc_url, node_rest, chain_id, keypair.address, keypair.valoper_address,
            cfg.poll_interval_seconds, state.last_scanned_block, price_source is not None,
        )

        cycle = Cycle(rest=rest, steem=steem, keypair=keypair, chain_id=chain_id, cfg=cfg, state_dir=state_dir)

        # Heartbeat: the "idle"/"waiting" logs below are debug-level (filtered
        # out by default), so a quiet cycle -- no new transfers to attest,
        # which is most cycles most of the time -- produces zero output at
        # all without this. Mirrors oracle/go/relayer/relayer.go's heartbeat.
        HEARTBEAT_EVERY = 10
        tick_count = 0

        while not stop.is_set():
            if stop.wait(cfg.poll_interval_seconds):
                break

            try:
                state = run_cycle(cycle, state, not_bonded_logged)
            except Exception:  # noqa: BLE001
                logger.exception("steem oracle cycle failed")

            if price_source is not None:
                try:
                    run_price_feeder_cycle(cycle, feeder, gas_prices, last_handled_period)
                except Exception:  # noqa: BLE001
                    logger.exception("price feeder cycle failed")

            tick_count += 1
            if tick_count % HEARTBEAT_EVERY == 0:
                logger.info("steem oracle heartbeat: last_scanned_block=%d", state.last_scanned_block)

        logger.info("steem oracle stopped")
    finally:
        rest.close()
        steem.close()
